components/script/dom/gpubuffer.rs

```diff
@@ -313,24 +317,19 @@ impl GPUBufferMethods for GPUBuffer {
             return Err(Error::Operation);
         }
 
-        unsafe extern "C" fn free_func(_contents: *mut c_void, free_user_data: *mut c_void) {
-            let _ = Rc::from_raw(free_user_data as _);
-        }
-
-        let array_buffer = unsafe {
-            NewExternalArrayBuffer(
-                *cx,
-                range_size as usize,
-                m_info.mapping.borrow_mut()[offset as usize..m_end as usize].as_mut_ptr() as _,
-                Some(free_func),
-                Rc::into_raw(m_info.mapping.clone()) as _,
-            )
-        };
+        let heap_typed_array = create_new_external_array_buffer::<ArrayBufferU8>(
+            cx,
+            Arc::clone(&m_info.mapping),
+            offset as usize,
+            range_size as usize,
+            m_end as usize,
+        );
 
+        let result = heap_typed_array.get_internal().map_err(|_| Error::JSFailed);
```

Comment:
Here we should also expect, see https://github.com/servo/servo/pull/31202/files#diff-caa3a7eda41967ceeede122072e749d9bda074ea1860d63c862817b61bc6baf0L333

Reply:
Just FYI, the one change where I said to return an error, is actually something I saw in the spec. Here the spec does not allow the operation to fail.

Reply:
Ah, perhaps in that case it is better to panic.

Reply:
Yes actually while we're at it we should update our IDL because the current ones in the spec make it immediately clear that those are not fallible methods(must have been a previous version). 

@Taym95 Can you pls do it? It shouldn't require much changes. Sorry to keep adding stuff!

See the spec: https://gpuweb.github.io/gpuweb/#gpubuffer

And ours: https://github.com/servo/servo/blob/ba42f7b6b121d7edc367cac22b60613da5487144/components/script/dom/webidls/WebGPU.webidl#L160

Reply:
Previous comment was incorrect, removed it.

Reply:
What we're doing here is in fact: https://webidl.spec.whatwg.org/#arraybuffer-create

Called at step 4 of https://gpuweb.github.io/gpuweb/#dom-gpubuffer-getmappedrange

Compare with step 2 which we handle above at https://github.com/servo/servo/pull/31202/files/ba42f7b6b121d7edc367cac22b60613da5487144#diff-caa3a7eda41967ceeede122072e749d9bda074ea1860d63c862817b61bc6baf0R299

Reply:
Okay. This sounds reasonable.

Reply:
Added expect